feat: MDS connections use mTLS

nolanleastin opened the pull request:
**Feature Gating**
The `GCE_METADATA_MTLS_MODE` environment variable is introduced, which can be set to strict, none, or default.

The `should_use_mds_mtls` function determines whether to use mTLS based on the environment variable and the existence of the certificate files.

**Using MDS mTLS**
A custom `MdsMtlsAdapter` for requests is implemented to handle the SSL context for mTLS.

The `create_session` function mounts an MdsMtlsAdapter into the request.Session

**MdsMtlsAdapter**
Loads MDS mTLS certificates from well-known location (https://docs.cloud.google.com/compute/docs/metadata/overview#https-mds-certificates).

If HTTPS request fails, MdsMtlsAdapter will fallback to HTTP. 

**Integrating with existing code**
compute_engine/_metadata.py:
- The metadata server URL construction is now dynamic, supporting both http and https schemes based on whether mTLS is enabled.
- ping and get functions are updated to use mTLS when it's enabled.

nolanleastin committed:
feat: MDS connections use mTLS

nolanleastin committed:
add custom root unit test to increase coverage

nolanleastin committed:
add unit test to increase coverage

nolanleastin commented on google/auth/compute_engine/_mtls.py:
```diff
@@ -0,0 +1,109 @@
+
+    def __init__(self, mds_mtls_config: MdsMtlsConfig, *args, **kwargs):
+        self.ssl_context = ssl.create_default_context()
+        self.ssl_context.load_verify_locations(cafile=mds_mtls_config.ca_cert_path)
```
ca_cert_path points to the `root.crt` -- this is the self-signed CA certificate, used to verify the server identity (i.e. MDS)

client_combined_cert_path points to a file called `client_creds.key` -- this has  the client certificate and private key.

nolanleastin committed:
add explanation to mtlsmds modes

nolanleastin commented on google/auth/compute_engine/_mtls.py:
```diff
@@ -0,0 +1,109 @@
+class MdsMtlsMode(enum.Enum):
+    """MDS mTLS mode."""
+
+    STRICT = "strict"
```
Done!

nolanleastin committed:
Update mds mtls certificate well-known locations

daniel-sanche commented on google/auth/compute_engine/_metadata.py:
```diff
@@ -102,6 +115,24 @@ def detect_gce_residency_linux():
+
+    Returns:
+        google.auth.transport.Request: Request
+            object to use.
```
nit: this could be a single line

daniel-sanche commented on google/auth/compute_engine/_mtls.py:
```diff
@@ -0,0 +1,135 @@
+
+
+def _get_mds_root_crt_path():
+    if os.name == "nt":
```
nit: consider making this special nt string into a constant

daniel-sanche reviewed (commented):
Nothing major, but left a few suggestions to consider

daniel-sanche commented on google/auth/compute_engine/_metadata.py:
```diff
@@ -115,6 +146,8 @@ def ping(request, timeout=_METADATA_DEFAULT_TIMEOUT, retry_count=3):
         bool: True if the metadata server is reachable, False otherwise.
     """
+    use_mtls = _mtls.should_use_mds_mtls()
+    request = _prepare_request_for_mds(request, use_mtls=use_mtls)
```
Isn't this a private file? Do users call these functions directly?

nolanleastin replied:
I have modified this. I no longer create a new request object!

daniel-sanche reviewed (changes requested):
Looking at this a bit closer, it looks like this change would [contradict the public API](https://github.com/googleapis/google-auth-library-python/blob/89a883824b437e9be0c97309ed36f35e5c071f11/google/oauth2/id_token.py#L255) as it currently stands. Custom request objects constructed and passed in by users would be bypassed after the next release, which could break existing code. I think we need to consider that more before merging 

Or let me know if I'm misunterstanding

nolanleastin committed:
modify contants

nolanleastin committed:
change the mds mtls implementation

1. now we do not create a new request. instead, create an mds mtls
   adapter and mount it on the request session.

2. added _validate_gce_mds_configured_environment, which ensures if we
   are using strict, that the host being contacted is default

3. fix unit tests and add new tests

nolanleastin committed:
add fallback to mds mtls

nolanleastin commented on google/auth/compute_engine/_mtls.py:
```diff
@@ -0,0 +1,135 @@
+            "C:\\", "ProgramData", "Google", "ComputeEngine", "mds-mtls-root.crt"
+        )
+    else:
+        return os.path.join("/", "run", "google-mds-mtls", "root.crt")
```
Done

nolanleastin commented on google/auth/compute_engine/_metadata.py:
```diff
@@ -102,6 +115,24 @@ def detect_gce_residency_linux():
+
+    Args:
+        request (google.auth.transport.Request): A callable used to make
+            HTTP requests.
```
done

nolanleastin committed:
add fallback for http error codes from mds

daniel-sanche commented on google/auth/compute_engine/_mtls.py:
```diff
@@ -0,0 +1,164 @@
+    )  # path to CA certificate
+    client_combined_cert_path: str = field(
+        default_factory=_get_mds_client_combined_cert_path
+    )  # path to file containing client certificate and key
```
these are marked as str, but it looks like they are paths now?

Are you running `nox -s mypy` checks? I'm not sure if they're configured to run in CI

nolanleastin committed:
Add more unit tests to address coverage requirements

nolanleastin committed:
update docstrings

nolanleastin committed:
update MdsMtlsConfig fields to be Path type

sai-sunder-s reviewed (commented):
This is a great addition to the library. The implementation of mTLS for the GCE metadata server is well-structured and thoroughly tested.

Here's a summary of my review:
- **Good separation of concerns:** The mTLS logic is nicely encapsulated in the new `_mtls.py` module.
- **Clear configuration:** The use of the `GCE_METADATA_MTLS_MODE` environment variable provides a clear and flexible way to configure the mTLS behavior.
- **Robustness:** The fallback mechanism in `default` mode is a good feature to ensure that the library can still function even if mTLS is not properly configured.
- **Thorough testing:** The addition of new test files and the extension of existing tests provide good coverage for the new functionality.

I have one minor suggestion for improvement in `google/auth/compute_engine/_metadata.py`, which I've added as a line comment.

Overall, this looks ready to merge.

sai-sunder-s commented on google/auth/compute_engine/_metadata.py:
```diff
@@ -102,6 +143,32 @@ def detect_gce_residency_linux():
+        request.session = requests.Session()
+    for host in _GCE_DEFAULT_MDS_HOSTS:
+        request.session.mount(f"https://{host}/", adapter)
+    return request
```
you are returning the same request object as the input. All changes are inside the object.

Do we really have to return?

sai-sunder-s reviewed (dismissed):
Overall LGTM. Left couple of comments.

nolanleastin commented on google/auth/compute_engine/_metadata.py:
```diff
@@ -24,31 +24,72 @@
+        if _GCE_METADATA_HOST != _GCE_DEFAULT_HOST:
+            # mTLS is only supported when connecting to the default metadata host.
+            # Raise an exception if we are in strict mode (which requires mTLS)
+            # but the metadata host has been overridden. (which means mTLS will fail)
```
This is a good point. I will make this edit.

nolanleastin committed:
address feedback

daniel-sanche reviewed (dismissed):
LGTM

nolanleastin committed:
minor modification to docstring
